### ISSUE TITLE
Add new federated learning method: average the gradients instead of averaging the weights 

### ISSUE THREAD

**Issue by RomainGoussault** [SubstraFoundation/distributed-learning-contributivity#177]
This method is called (on wikipedia), [Federated Stochastic Gradient Descent (FedSGD)](https://en.wikipedia.org/wiki/Federated_learning#Federated_Stochastic_Gradient_Descent_(FedSGD)).

I guess this is the same as the Large Batch method described [here ](https://medium.com/@Edgify/distributed-training-on-edge-devices-large-batch-vs-federated-learning-b61f70b1c734)by the Edigfy company.

**Comment by Thomas-Galtier** [comment]
I've just noticed that averaging the gradient is in fact equivalent to averaging the weights. So I wonders if implementing this approach is really worth it. What do you think?

**Comment by RomainGoussault** [comment]
The [FedAvg paper](https://arxiv.org/pdf/1602.05629.pdf) provides some comparison between FedAvg and FedSgd

**Comment by Thomas-Galtier** [comment]
Ok so the FedAvg method is better.

**Comment by bowni** [comment]
@arthurPignet to make the link with what you started working on today

**Comment by bowni** [comment]
See also #173 which was marked duplicate, but contains interesting comments

**Comment by arthurPignet** [comment]
I made two implementations on the library of what I denote gradfusion.

It appears that this technique showns two main advantages on fedavg.

As we are using aggregating the gradients, at the end we use only one optimizer, so we can use state of the art optimizer, which is not the case when using fedavg. It dramatically improve the convergence of the model through the epochs. 

Secondly, we are not using get/set_weights methods, so the model is only modified by the optimizer. It increases the speed of training, and allow the use of tf function for a whole epoch. It makes possible an implementation which is equivalent in term of speed with the keras' .fit() method. 

The drawback is that the gradient_pass_per_update_count is currently fixed to one. So it increases the number of communication required between server and partners. 
 
One could think of an hybrid method, where several pass_per_update_count can be done locally, and the gradients coming from these updates are summed, and the result sent to the server. I don't know how the optimizer server will react to these summed-over-local-batch gradients.
